objgraph.py

```diff
@@ -355,6 +356,91 @@ def show_growth(limit=10, peak_stats=None, shortnames=True, file=None,
             file.write('%-*s%9d %+9d\n' % (width, name, count, delta))
 
 
+def get_ids(skip_update=False, limit=10, sortby='deltas',
+            OLD_IDS=collections.defaultdict(set),
+            CURRENT_IDS=collections.defaultdict(set),
+            NEW_IDS=collections.defaultdict(set)):
+    """Show the increase in object counts since last call to this function
+    and returns the memory address ids for old, current, and new objects.
+
+    Returns: [OLD_IDS, CURRENT_IDS, NEW_IDS]
+
+    OLD_IDS: a dictionary which stores sets of ids under the keys of object
+    type. These are the objects that were stored the last time the function was
+    called.
+
+    CURRENT_IDS: a dictionary which stores sets of ids under the keys of object
+    type. These are the objects that are stored when the function is called
+    now.
+
+    NEW_IDS: a dictionary which stores sets of ids under the keys of object
+    type. These are the objects that were created between the last time this
+    function was called and when it was called now.
+
+    sortby: This is the column that you want to sort by in descending order.
+    Possible values are: 'old', 'current', 'new', 'deltas'
+
+    limit: the maximum number of rows that you want to print data for
+
+    if ``skip_update`` is True, the sets of [OLD_IDS, CURRENT_IDS, NEW_IDS]
+    will be returned from when the function was last run without examining the
+    objects currently iin memory.
+
+    The caveats documented in :func:`growth` apply.
+
+    Example:
+
+        >>> get_ids()
+        ======================================================================
+        Type                    Old_ids  Current_ids      New_ids Count_Deltas
+        ======================================================================
+        weakref                    3807         3864          +57          +57
+        dict                       6892         6947          +73          +55
+        frame                        34           70          +53          +36
+        ...
```

Comment:
Ah, you have some tests!  Sorry I missed them on the first glance.

Reply:
API-wise I'm not very happy about functions that both print and return values.

Despite having read the description, I don't quite get how this one differs from `growth()`/`show_growth()`.

Also, the name (`get_ids`) is too generic, and doesn't describe what the function does.

(I'm afraid my criticism is not very constructive at this point.  I'll have to think about this more.)

Reply:
Ah, so this function also observes (or tries to observe) the object churn!  Interesting.

Given how object IDs can be reused by Python, are the numbers accurate?

Reply:
Suggestion for the function name: `get_new_ids()`.  It could also return just the NEW_IDs, for simplicity's sake.

I'm thinking the three parameters for tracking state are a bit unwieldy (e.g. you cannot do `OLD_IDS = CURRENT_IDS; CURRENT_IDS = defaultdict(set)`), so how about one `_state={}` parameter that you initialize like this:

```python
def get_new_ids(..., _state={}):
    ...
    _state['old'] = old = _state.get('current', defaultdict(set))
    _state['current'] = current = defaultdict(set)
    _state['new'] = new = defaultdict(set)
    ...
    return new
```

Reply:
> Given how object IDs can be reused by Python, are the numbers accurate?

I'm convinced now that given the purpose of this function that doesn't matter.  Objects that were freed and reallocated at the same memory address (while maintaining the same type name) aren't contributing to memory leaks.

Reply:
If we change the code to:
_state['old'] = old = _state.get('current', defaultdict(set))
Then the id of _state['old'] will change to the id of _state['current'].
This would also require a new dict be created every time for _state['current'] with new sets under every class_name.

I want to minimize the creation of new ids to store this info, so I prefer the old_dict[class_name].clear(), old_dict[class_name].update(current_dict[class_name])
current_dict[class_name].clear(), current_dict[class_name].add(new_id)
 pattern that way we re-use dictionaries and sets that are already reserved in memory.

Reply:
That is a very good point!